### PR TITLE
feat: add support to passing bootnodes as argument through the cli

### LINKED ISSUE #20: Add support to passing `bootnodes` as argument to the cli.
The client should support passing bootnodes through the CLI. These should be comma-separated strings, i.e

`bootnodes=enode://4e5e92199ee224a01932a377160aa432f31d0b351f84ab413a8e0a42f4f36476f8fb1cbe914af0d9aef0d51665c214cf653c651c4bbd9d5550a934f241f1682b@138.197.51.181:30303, enode://143e11fb766781d22d92a2e33f8f104cddae4411a122295ed1fdb6638de96a6ce65f5b7c964ba3763bba27961738fef7d3ecc739268f3e5e771fb4c87b6234ba@146.190.1.103:30303`

We should be able to parse these as Node Records.

Ethereum package will use to create a localnet: https://github.com/lambdaclass/ethereum-package/pull/1/files#diff-79bf8330a2e72edba847d9ca9de634c4339dc13afb11636b13e0a4775119dcbbR227-R234

When using a public network, we should have the values hardcoded in code/config.

These will be used for P2P discovery

### PR DESCRIPTION
**Pull request by ricomateo**
**Motivation**

According to the issue this will be used for P2P discovery

**Description**

This PR adds support to passing bootnodes as argument through the cli

Closes #20

### CODE REVIEW

File: ethrex/src/cli.rs [hunk 1]
```diff
@@ -32,11 +32,46 @@ pub fn cli() -> Command {
                 .value_name("PORT")
                 .action(ArgAction::Set),
         )
+        .arg(
+            Arg::new("addr")
```
Comment (mpaulucci) [inline ethrex/src/cli.rs:36]: it is not clear to me what this address means? Maybe we should prefix it like we do with the other ones, like authrpc, discovery, etc..

File: ethrex/src/main.rs [hunk 2]
```diff
@@ -25,17 +32,68 @@ fn main() {
...
     let genesis_file = std::fs::File::open(genesis_file_path).expect("Failed to open genesis file");
     let genesis_reader = BufReader::new(genesis_file);
     serde_json::from_reader(genesis_reader).expect("Failed to read genesis file")
 }
+
+fn parse_socket_addr(addr: &str, port: &str) -> Result<SocketAddr, AddrParseError> {
+    format!("{addr}:{port}").parse()
+}
+
+fn parse_bootnode(input: &str) -> BootNode {
```
Comment (mpaulucci) [inline ethrex/src/main.rs:82]: should we add a From implementation inside BootNode and move this logic there?

File: ethrex/src/cli.rs (hunk 1 above)
Comment (ricomateo) [reply ethrex/src/cli.rs:36]: I'm not sure what prefix should I put on it, maybe @MegaRedHand can give us some insight

File: ethrex/src/main.rs (hunk 2 above)
Comment (ricomateo) [reply ethrex/src/main.rs:82]: Definitely. 
What do you think about storing the `node_id` in a `Vec<u8>`? Should I keep it that way or change it to a `U512`? or maybe to another type?

Comment (mpaulucci) [reply ethrex/src/main.rs]: U512 sounds like a good option

File: crates/net/Cargo.toml [hunk 3]
```diff
@@ -8,3 +8,4 @@ edition = "2021"
 [dependencies]
 tracing.workspace = true
 tokio.workspace = true
+ethereum-types = "0.14.1"
```
Comment (MegaRedHand) [inline crates/net/Cargo.toml:11]: We shouldn't depend on `ethereum-types`. For that we have the `core` crate.

File: ethrex/src/cli.rs (hunk 1 above)
Comment (MegaRedHand) [reply ethrex/src/cli.rs]: The naming was taken from geth/reth. It's used for both discovery and rlpx (`discovery.addr` only overrides the discovery part).

File: ethrex/src/main.rs (hunk 2 above)
Comment (MegaRedHand) [reply ethrex/src/main.rs]: Let's use `H512` since it's a hash

File: ethrex/src/cli.rs (hunk 1 above)
Comment (mpaulucci) [reply ethrex/src/cli.rs]: I like `p2p.addr`

File: crates/net/src/types/bootnode.rs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+impl FromStr for BootNode {
+    type Err = ParseIntError;
+    /// Takes a str with the format "enode://<node ID>@<IP address>:<port>" and
+    /// parses it to a BootNode
+    fn from_str(input: &str) -> Result<BootNode, ParseIntError> {
+        // TODO: error handling
+        let node_id_as_bytes = decode_hex(&input[8..136])?;
+        let node_id = H512::from_slice(&node_id_as_bytes);
```
Comment (MegaRedHand) [inline crates/net/src/types/bootnode.rs:17]: We can use `H512::from_str` for this

File: crates/net/src/types/bootnode.rs [hunk 5]
```diff
@@ -0,0 +1,33 @@
+use ethrex_core::H512;
+use std::{net::SocketAddr, num::ParseIntError, str::FromStr};
+
+#[derive(Debug, Clone)]
+pub struct BootNode {
+    pub node_id: H512,
```
Comment (MegaRedHand) [inline crates/net/src/types/bootnode.rs:6]: This should be `H256` since the ID is a hash.

File: crates/net/src/types/bootnode.rs [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+        })
+    }
+}
+
+pub fn decode_hex(s: &str) -> Result<Vec<u8>, ParseIntError> {
+    (0..s.len())
+        .step_by(2)
+        .map(|i| u8::from_str_radix(&s[i..i + 2], 16))
+        .collect()
+}
```
Comment (MegaRedHand) [inline crates/net/src/types/bootnode.rs:33]: Can we add some tests?
For example, deserializing this EF bootnode entry: `enode://d860a01f9722d78051619d1e2351aba3f43f943f6f00718d1b9baa4101932a1f5011f16bb2b1bb35db20d6fe28fa0bf09636d26a87d31de9ec6203eeedb1f666@18.138.108.67:30303`

**Review by MegaRedHand — CHANGES_REQUESTED** [review]

File: crates/net/src/types/bootnode.rs (hunk 5 above)
Comment (MegaRedHand) [reply crates/net/src/types/bootnode.rs:6]: According to the documentation it's a pubkey (probably secp256k1), of 64 bytes (512 bits)

**Review by MegaRedHand — APPROVED** [review]
LGTM

**Pull request merged**